### PR TITLE
WIP: providers: Add cognito as a provider

### PR DESCRIPTION
**Pull request by jpc2350**
Hi, I've been using AWS Cognito a lot recently as an Oauth provider and thought it would be worth sharing.

First time contributing so please let me know if there is anything that isn't up to scratch - I think i've followed everything that was in the contrib README.

Thanks

### CODE REVIEW

File: flask_dance/contrib/cognito.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    :returns: A :ref:`blueprint <flask:blueprints>` to attach to your Flask app.
+    """
+    scope = scope or ["openid", "profile"]
+
+    cognito_bp = OAuth2ConsumerBlueprint(
+        "cognito",
+        __name__,
+        client_id=client_id,
+        client_secret=client_secret,
+        base_url="https://{cognito_suffix}.auth.eu-west-1.amazoncognito.com".format(
```
[the remark this answers is not in the document]
Comment (jpc2350) [reply flask_dance/contrib/cognito.py]: avoided by requiring an AWS_REGION parameter

File: flask_dance/contrib/cognito.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            cognito_suffix=cognito_domain_suffix
+        ),
+        authorization_url="https://{cognito_suffix}.auth.eu-west-1.amazoncognito.com/oauth2/authorize".format(
+            cognito_suffix=cognito_domain_suffix
+        ),
+        token_url="https://{cognito_suffix}.auth.eu-west-1.amazoncognito.com/oauth2/token".format(
+            cognito_suffix=cognito_domain_suffix
+        ),
+        # scope=["openid", "profile"],
```
[the remark this answers is not in the document]
Comment (jpc2350) [reply flask_dance/contrib/cognito.py]: Removed

File: flask_dance/contrib/cognito.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    Make a blueprint for authenticating with Cognito using OAuth 2. This requires
+    a client ID, client secret and URLS from Cognito. You should either pass them to
+    this constructor, or make sure that your Flask application config defines
+    them, using the variables :envvar:`COGNITO_OAUTH_CLIENT_ID` and
+    :envvar:`COGNITO_OAUTH_CLIENT_SECRET`.
+    Args:
+        cognito_domain_suffix (str): The Cognito Domain Suffix for your cognito user pool
+        client_id (str): The client ID for your application on Cognito.
+        client_secret (str): The client secret for your application on Cognito
+        scope (str, optional): comma-separated list of scopes for the OAuth token. Defaults to ["openid", "profile"]
```
[the remark this answers is not in the document]
Comment (jpc2350) [reply flask_dance/contrib/cognito.py]: Done

File: flask_dance/contrib/cognito.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    cognito_domain_suffix=None,
+):
+    """
+    Make a blueprint for authenticating with Cognito using OAuth 2. This requires
+    a client ID, client secret and URLS from Cognito. You should either pass them to
+    this constructor, or make sure that your Flask application config defines
+    them, using the variables :envvar:`COGNITO_OAUTH_CLIENT_ID` and
+    :envvar:`COGNITO_OAUTH_CLIENT_SECRET`.
+    Args:
+        cognito_domain_suffix (str): The Cognito Domain Suffix for your cognito user pool
```
[the remark this answers is not in the document]
Comment (jpc2350) [reply flask_dance/contrib/cognito.py]: I've changed this to Domain Name. 

You have to use your own domain for cognito (or the one that you've set cognito up with). Unfortunately, there is no default domain for cognito.

**Comment by jpc2350** [comment]
Thanks! I've addressed the requested changes and added a Cognito section in docs/providers.rst for this. Hopefully, the latest commits should 

Yes, I have been using this integration on an internal production system for the last two months. (And Tox tested too).

**Comment by singingwolfboy** [comment]
Based on the AWS documentation, [it appears that Cognito supports using your own domain](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-add-custom-domain.html). Am I misunderstanding this? If Cognito supports it but this Flask-Dance configuration does not, that's OK as long as we document that. (But it would be better if this configuration supported it.)

It also appears that the `domain_name` argument is required. Can you add a check in the code so that if it is unset, the provider throws an informative error message? We'll need an automated test to cover that, too.

**Comment by jpc2350** [comment]
@singingwolfboy I hadn't seen that. Let me have a look and re-factor the code to support this feature.

Will add that too.

Thanks

**Comment by trickidicki** [comment]
The AWS domains look correct. Can I suggest to add another variable to the blueprint for the logout URL since I spent some time trying to find it:

```
cognito_bp.logout_url = "https://{domain_name}.auth.{aws_region}.amazoncognito.com/logout".format(
        userpool=userpool,region=region)
```

**Comment by singingwolfboy** [comment]
@jpc2350: instead of asking the developer to provide `domain_name` and `aws_region` to interpolate into `"https://{domain_name}.auth.{aws_region}.amazoncognito.com"`, why not just ask the user to provide the whole domain? That would allow for both `amazoncognito.com` domains and custom domains.

**Comment by singingwolfboy** [comment]
Hi @jpc2350, are you planning to continue working on this, or should I close the pull request?

**Comment by jpc2350** [comment]
I'm planning to continue, but haven't had time at present - I'll close for now, and open a new one when I'm ready

**Pull request closed without merging**